quinn-proto/src/connection.rs

```diff
@@ -1031,58 +1037,87 @@ impl Connection {
         }
 
         trace!(ctx.log, "connection got packet"; "connection" => %self.loc_cid, "len" => packet.payload.len());
-        let was_closed = self.state.as_ref().unwrap().is_closed();
+        let (prev_state, was_handshake) = match self.state.take().unwrap() {
+            State::Handshake(mut state) => {
+                if !state.rem_cid_set {
+                    match packet.header {
+                        Header::Long {
+                            src_cid: rem_cid, ..
+                        } => {
+                            trace!(ctx.log, "got remote connection id"; "connection" => %self.loc_cid, "rem_cid" => %rem_cid);
+                            self.rem_cid = rem_cid;
+                            state.rem_cid_set = true;
+                        }
+                        _ => {}
+                    }
+                }
+                (State::Handshake(state), true)
+            }
+            state => (state, false),
+        };
+        let was_closed = prev_state.is_closed();
 
-        // State transitions
-        let prev_state = self.state.take().unwrap();
-        let was_handshake = match prev_state {
-            State::Handshake(_) => true,
-            _ => false,
+        let number = match self.decrypt_packet(was_handshake, &mut packet) {
+            Ok(number) => Some(number),
+            Err(Some(e)) => {
+                warn!(ctx.log, "got illegal packet"; "connection" => %self.loc_cid, "reason" => %e);
+                None
+            }
+            Err(None) => {
+                debug!(ctx.log, "failed to authenticate packet"; "connection" => %self.loc_cid);
+                None
+            }
         };
 
-        let state = match self.handle_connected_inner(ctx, now, remote, packet, prev_state) {
-            Ok(state) => state,
-            Err(conn_err) => {
-                ctx.events.push_back((
-                    self.handle,
-                    Event::ConnectionLost {
-                        reason: conn_err.clone(),
-                    },
-                ));
+        // State transitions
+        let state = if let Some(number) = number {
+            self.on_packet_authenticated(ctx, now, number);
```

Comment:
This should only be called when handshaking or established. Might also be clearer if positioned in the match on `decrypt_packet`'s return value, since that's when we deem the packet authenticated.

Reply:
See also in d26777df2414958acb5b479de4345a76acd5d24b.